pandas/tests/groupby/test_categorical.py

```diff
@@ -453,6 +453,44 @@ def test_dataframe_categorical_with_nan(observed):
     tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize("ordered", [True, False])
+@pytest.mark.parametrize("observed", [True, False])
+@pytest.mark.parametrize("sort", [True, False])
+def test_dataframe_categorical_ordered_observed_sort(ordered, observed, sort):
+    # GH 25871: Fix groupby sorting on ordered Categoricals
+    # Build a dataframe with cat having one unobserved category ('AWOL'),
+    # and a Series with identical values
+    cat = pd.Categorical(['d', 'a', 'b', 'a', 'd', 'b'],
+                         categories=['a', 'b', 'AWOL', 'd'],
+                         ordered=ordered)
+    val = pd.Series(['d', 'a', 'b', 'a', 'd', 'b'])
+    df = pd.DataFrame({'cat': cat, 'val': val})
+
+    # aggregate on the Categorical
+    result = (df.groupby('cat', observed=observed, sort=sort)['val']
+                .aggregate('first'))
+
+    # If ordering works, we expect index labels equal to aggregation results,
+    # except for 'observed=False': index contains 'AWOL' and aggregation None
+    label = pd.Series(result.index.array, dtype='object')
+    aggr = pd.Series(result.array)
+    if not observed:
+        aggr[aggr.isna()] = 'AWOL'
+    tm.assert_series_equal(label, aggr)
```

Comment:
ideally would like this to simply create an expected result and compare the entire dataframe with assert_frame_equal. this might not be trivial to do though.

Reply:
I had an earlier version, with only four parameter combinations, where I generated four expected DataFrames. This got old though, especially as the parameters cause subtle differences in the ordering of codes vs. categories in the result's Categorial index, even for the (visibly) same result data frame. 
In the end, I argued that for this issue, I'm really only interested in the result order anyway, rather than the internal details of the DataFrame. What's more, future refactoring might change those internals while leaving the result correctly ordered. 

Still, if you wish me to do it I will generate 8 expected result DataFrames to compare against. Your thoughts?

Reply:
i am kind of ok with the labels checking as you had done, though ideally would have hard coded expected results (which are then easy to see & logically interpret). even for the lables I would like to hard code them, or at least make them obvious.

Reply:
Would you be OK with the check remaining the same, but the assertion reporting 
```
AssertionError: Labels and aggregation results not consistently sorted
for (ordered=True, observed=True, sort=True)
Result:
label
d    a
a    b
b    d
Name: val, dtype: object
```
in case of failure?
That should make things more obvious.

Reply:
I've gone ahead and committed this, feel free to request further changes.